### PR TITLE
New ephemeral objects middleware and memcache extension

### PR DESCRIPTION
**Pull request by murlock**
Import Ephemeral object and Memcache middleware

### CODE REVIEW

File: swift/common/middleware/ephemeral_object.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+       Ephemeral Objects can be listed with a dedicated parameter
+         oio.list_mpu is implemented at this time
+    """
+
+    def __init__(self, app, conf, redis_conf=None, **kwargs):
+        self.app = app
+        # legacy parameters
+        self.account_first = kwargs.get('account_first', True)
+        self.swift3_compat = kwargs.get('swift3_compat', True)
+        self.strip_v1 = kwargs.get('strip_v1', True)
```
Comment (AymericDu) [inline swift/common/middleware/ephemeral_object.py]: We should use `config_true_value`

File: swift/common/middleware/ephemeral_object.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+        return account, container, obj
+
+    def _create_entry(self, keys, path, headers):
+        try:
+            # key = self.mpu_key(account, container)
+            # TODO(mbo): PY3: should we force encoding to bytes
+            self.conn.hset(keys[1], path, json.dumps(headers))
+
+            # key = self.key(account, container)
+            self.conn.zset(keys[0], path)
```
Comment (AymericDu) [inline swift/common/middleware/ephemeral_object.py]: We should use a transaction

File: swift/common/middleware/ephemeral_object.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+        except Exception as e:
+            LOG.error("%s: failed to create key %s (%s)", SWIFT_SOURCE,
+                      ':'.join([keys[0], path]), str(e))
+            raise HTTPInternalServerError()
+
+    def _remove_entry(self, keys, path):
+        # key = self.key(account, container)
+        self.conn.zdel(keys[0], path)
+        self.conn.hdel(keys[1], path)
```
Comment (AymericDu) [inline swift/common/middleware/ephemeral_object.py]: We should use a transaction

File: swift/common/middleware/ephemeral_object.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+    def _handle_mpu_object(self, req, start_response, account, container,
+                           path):
+        """Handler uploadId requests on memory backend.
+           Return tuple (response, fallback_needed)
+        """
+        lkey, hkey = self.keys(account, container)
+        if req.method in ('GET', 'HEAD'):
+            is_available = self.conn.hexists(hkey, path)
+            if is_available:
+                res = self.conn.hget(hkey, path)
```
Comment (AymericDu) [inline swift/common/middleware/ephemeral_object.py]:
```suggestion
            res = self.conn.hget(hkey, path)
            if res is not None:
```

File: swift/common/middleware/ephemeral_object.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        req = Request(env)
+        account, container, obj = self._extract_path(req.path_info)
+
+        if container is None or '+segments' not in container:
+            return self.app(env, start_response)
+
+        # if obj and prefix are None with container+segments, we want the
+        # normal listing because it is the list-multipart-uploads operation
+        # reserved for MPU uploadId
+        if req.environ.get('oio.list_mpu'):
```
Comment (AymericDu) [inline swift/common/middleware/ephemeral_object.py]: We should use only one environment variable : `oio.ephemeral_object`
We can know if it is a listing thanks to the method used and to the presence of an object in the path.

File: swift/common/middleware/ephemeral_object.py [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+            if not limit:
+                limit = DEFAULT_LIMIT
+            else:
+                limit = int(limit[0])
+            must_recurse = req.method == 'GET' and 'delimiter' not in qs
+
+            return self._build_mpu_listing(start_response, env,
+                                           account, container, prefix,
+                                           limit=limit, marker=marker,
+                                           recursive=must_recurse)
```
Comment (AymericDu) [inline swift/common/middleware/ephemeral_object.py]: To be more generic, we could do a classic listing and insert ephemeral objects in this listing.

File: swift/common/middleware/ephemeral_object.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+# For each UploadID, two redis keys will be created:
+# - one entry in zset MPU_LIST:{ACCOUNT}:{CONTAINER/BUCKET}: it will be used
+# to check if UploadId is still present and to support list-multipart-uploads
+# - one key as MPU_HDRS:{ACCOUNT}:{CONTAINER/BUCKET}:{UPLOADID} to hold headers
+LST_PREFIX = 'EO_LIST:'
+HDR_PREFIX = 'EO_HDRS:'
+
+# When iterating in Redis with non recursive listing, we force next
+# zrangebylex to skip already used prefix
+FAKE_MARKER = "\xef\xbf\xbd"
```
Comment (AymericDu) [inline swift/common/middleware/ephemeral_object.py]: In python3, I think it should be `\ufffd`.

Comment (fvennetier) [reply swift/common/middleware/ephemeral_object.py]: `u"\ufffd"` works for both Python 2 and 3.

File: swift/common/middleware/ephemeral_object.py (hunk 1 above)
Comment (fvennetier) [reply swift/common/middleware/ephemeral_object.py]: Done.

File: swift/common/middleware/ephemeral_object.py (hunk 4 above)
Comment (fvennetier) [reply swift/common/middleware/ephemeral_object.py]: Done.

File: swift/common/middleware/ephemeral_object.py (hunk 6 above)
Comment (fvennetier) [reply swift/common/middleware/ephemeral_object.py]: This sounds reasonable.

File: swift/common/middleware/ephemeral_object.py (hunk 1 above)
Comment (fvennetier) [reply swift/common/middleware/ephemeral_object.py]: Actually I got rid of these parameters.

File: swift/common/middleware/ephemeral_objects.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+# to check if UploadId is still present and to support list-multipart-uploads
+# - one key as MPU_HDRS:{ACCOUNT}:{CONTAINER/BUCKET}:{UPLOADID} to hold headers
+LST_PREFIX = 'EO_LIST:'
+HDR_PREFIX = 'EO_HDRS:'
+
+# HACK: When iterating in Redis with non recursive listing, we force the next
+# zrangebylex iteration to skip already known prefixes. This is done by
+# suffixing the last known element with the "unicode replacement character",
+# which is the last valid character.
+END_MARKER = u"\ufffd"
```
Comment (AymericDu) [inline swift/common/middleware/ephemeral_objects.py]: From Unicode 6.0, the last code point is `\U0010ffff`.

Comment (fvennetier) [reply swift/common/middleware/ephemeral_objects.py]: And the last **valid** character seems to be `\U0010fffd`.

File: swift/common/middleware/ephemeral_object.py (hunk 3 above)
Comment (fvennetier) [reply swift/common/middleware/ephemeral_object.py]: Done.

File: swift/common/middleware/ephemeral_object.py (hunk 2 above)
Comment (fvennetier) [reply swift/common/middleware/ephemeral_object.py]: Done.

**Review by fvennetier — APPROVED** [review]

**Review by AymericDu — APPROVED** [review]

**Pull request merged**